kong/runloop/plugins_iterator.lua

```diff
@@ -258,52 +257,95 @@ local function load_configuration_through_combos(ctx, combos, plugin)
 end
 
 
-local function get_next(self)
-  local i = self.i + 1
+local function get_next_with_ctx(ctx, loaded, phases, combos, map)
+  local i = 1
+  return function()
+    local cfg
+    local plugin = loaded[i]
+    while plugin do
+      local name = plugin.name
+      if map[name] then
+        local plugins = ctx.plugins
+
+        local cmbs = combos[name]
+        if cmbs then
+          cfg = load_configuration_through_combos(ctx, cmbs, plugin)
+          if cfg then
+            plugins[name] = cfg
+            if plugin.handler.response and plugin.handler.response ~= BasePlugin.response then
+              ctx.buffered_proxying = true
+            end
+          end
+        end
 
-  local plugin = self.loaded[i]
-  if not plugin then
-    return nil
-  end
+        if phases and phases[name] then
+          cfg = plugins[name]
+          if cfg then
+            break
+          end
+        end
+      end
 
-  self.i = i
+      i = i + 1
+      plugin = loaded[i]
+    end
 
-  local name = plugin.name
-  if not self.ctx then
-    if self.phases[name] then
-      return plugin
+    if plugin then
+      i = i + 1
     end
 
-    return get_next(self)
+    return plugin, cfg
   end
+end
 
-  if not self.map[name] then
-    return get_next(self)
-  end
 
-  local ctx = self.ctx
-  local plugins = ctx.plugins
-
-  if self.configure then
-    local combos = self.combos[name]
-    if combos then
-      local cfg = load_configuration_through_combos(ctx, combos, plugin)
-      if cfg then
-        plugins[name] = cfg
-        if plugin.handler.response and plugin.handler.response ~= BasePlugin.response then
-          ctx.buffered_proxying = true
+local function get_next_with_plugins(plugins, loaded, phases, map)
+  local i = 1
+  return function()
+    local cfg
+    local plugin = loaded[i]
+    while plugin do
+      local name = plugin.name
+      if map[name] then
+        if phases and phases[name] then
+          cfg = plugins[name]
+          if cfg then
+            break
+          end
         end
       end
+
+      i = i + 1
+      plugin = loaded[i]
     end
-  end
 
-  if self.phases[name] and plugins[name] then
-    return plugin, plugins[name]
+    if plugin then
+      i = i + 1
+    end
```

Comment:
when `plugin` is falsy, the iteration has ended, so there's no harm to increment; i'd think the condition itself can be as heavy as the increment.   did it make difference in the benchmarks?